### PR TITLE
feat: New LMS Config Workflow Skeleton + Typescript

### PR DESCRIPTION
**Pull request by marlonkeating**
[Jira Ticket](https://2u-internal.atlassian.net/browse/ENT-6592)

# For all changes

- [ ] Ensure adequate tests are in place (or reviewed existing tests cover changes)

# Only if submitting a visual change

- [ ] Ensure to attach screenshots
- [ ] Ensure to have UX team confirm screenshots

# TODOs

- [x] Complete unit tests
- [x] Prevent user entering duplicate canvas config display name

### CODE REVIEW

File: src/components/forms/FormWorkflow.tsx [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+              ))}
+            </Stepper>
+          )}
+          <span>
+            <Form>
+              {step && FormComponent && (
+                <>
+                  <FormComponent {...formWorkflowConfig.configProps} />
+                  <span className="d-flex">
+                    {/* TODO: Fix typescript issue with Paragon Button */}
```
Comment (marlonkeating) [inline src/components/forms/FormWorkflow.tsx:97]: We'll either need to find type definitions from the Paragon library Button component, or include our own.
image

Comment (adamstankiewicz) [reply src/components/forms/FormWorkflow.tsx:97]: Feel free to make an upstream Paragon contribution to add types to the `Button` component! The only existing component we've cut over to TypeScript thus far is `Bubble`, which defines an `interface` for the `BubbleProps`. While `BubbleProps` is exported in `./src/Bubble/index.tsx`, it's not exported by `./src/index.js`; without the `interface` exported there, would an MFE be able to pick up the types for the component? For example, if you define an `interface` for `Button`, does that interface need to be exported by Paragon's `./src/index.js`?

File: src/components/forms/ValidatedFormControl.tsx [hunk 2]
```diff
@@ -0,0 +1,53 @@
+import React from "react";
+import omit from "lodash/omit";
+
+import { Form } from "@edx/paragon";
+
+import { setFormFieldAction } from "./data/actions";
+import { useFormContext } from "./FormContext";
+
+// TODO: Add Form.Control props.  Does Paragon export?
```
Comment (adamstankiewicz) [inline src/components/forms/ValidatedFormControl.tsx:9]: It does not currently :)

File: src/components/settings/SettingsLMSTab/LMSConfigs/Canvas/CanvasConfig.tsx [hunk 3]
```diff
@@ -0,0 +1,126 @@
+import { SyntheticEvent } from "react";
+import isEmpty from "lodash/isEmpty";
+
+import handleErrors from "../../../utils";
+import LmsApiService from "../../../../../data/services/LmsApiService";
+import { snakeCaseDict } from "../../../../../utils";
```
Comment (adamstankiewicz) [inline src/components/settings/SettingsLMSTab/LMSConfigs/Canvas/CanvasConfig.tsx:6]: nit: `@edx/frontend-platform/utils` also exports a `snakeCaseObject` utility function, similar to its `camelCaseObject` function.

File: src/components/forms/FormWorkflow.tsx [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+  dispatch,
+}: FormWorkflowProps<FormData>) {
+  const {
+    formFields,
+    currentStep: step,
+    hasErrors,
+    isEdited,
+  }: FormContext = useFormContext();
+  const [errorIsOpen, openError, closeError] = useToggle(false);
+  const [savedChangesModalIsOpen, openModal, closeModal] = useToggle(false);
```
Comment (kiram15) [inline src/components/forms/FormWorkflow.tsx:61]: nit: I think the open and closeModal vars should be better named like openSaveModal, closeSaveModal

File: src/components/forms/FormWorkflow.tsx [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+    }
+  };
+
+  const stepBody = (step: FormWorkflowStep<FormData>) => {
+    const FormComponent: DynamicComponent = step?.formComponent;
+    return (
+      <Stepper.Step eventKey={step.index.toString()} title={step.stepName}>
+        <span>
+          {step && step?.formComponent && (
+            <>
```
Comment (kiram15) [inline src/components/forms/FormWorkflow.tsx:88]: Why is the empty container needed around this? And also the span above it?

File: src/components/forms/FormWorkflow.tsx [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+      </Stepper.Step>
+    );
+  };
+
+  const stepActionRow = (step: FormWorkflowStep<FormData>) => {
+    return (
+      <Stepper.ActionRow eventKey={step.index.toString()}>
+        <span className="d-flex">
+          {/* TODO: Help Link */}
+          {/* TODO: Fix typescript issue with Paragon Button */}
```
Comment (kiram15) [inline src/components/forms/FormWorkflow.tsx:102]: What was the issue?

File: src/components/forms/data/reducer.ts [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+      ?.map((validation: FormFieldValidation) => [
+        validation.formFieldId,
+        state.formFields && validation.validator(state.formFields),
+      ])
+      .filter((err) => !!err[1]);
+    if (!isEmpty(errors)) {
+      // Generate index of errors
+      errorState = {
+        hasErrors: true,
+        errorMap: groupBy(errors, (error) => error[0]),
```
Comment (kiram15) [inline src/components/forms/data/reducer.ts:53]: maybe add a small comment on this line to clarify what you're accomplishing

File: src/components/settings/SettingsLMSTab/LMSConfigs/Canvas/CanvasConfig.tsx [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+  clientId: string;
+  clientSecret: string;
+  id: string;
+  active: boolean;
+  uuid: string;
+};
+
+// TODO: Can we generate this dynamically?
+// https://www.typescriptlang.org/docs/handbook/2/mapped-types.html
+export type CanvasConfigSnakeCase = {
```
Comment (kiram15) [inline src/components/settings/SettingsLMSTab/LMSConfigs/Canvas/CanvasConfig.tsx:28]: Remove and use snakecasedict from platform!

File: src/components/settings/SettingsLMSTab/LMSConfigs/Canvas/CanvasConfig.tsx [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+  // TODO: Fix handleSubmit
+  const steps: FormWorkflowStep<CanvasConfigCamelCase>[] = [
+    {
+      index: 0,
+      formComponent: CanvasConfigAuthorizePage,
+      validations,
+      stepName: "Authorize",
+      saveChanges,
+      nextButtonConfig: {
+        buttonText: "Authorize",
```
Comment (kiram15) [inline src/components/settings/SettingsLMSTab/LMSConfigs/Canvas/CanvasConfig.tsx:124]: Should the text be constants somewhere?

File: src/components/settings/SettingsLMSTab/LMSConfigs/Canvas/CanvasConfigAuthorizePage.tsx [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+      return error && "Please enter a valid URL";
+    },
+  },
+  {
+    formFieldId: formFieldNames.DISPLAY_NAME,
+    validator: (fields) => {
+      // TODO: Check for duplicate display names
+      const displayName = fields[formFieldNames.DISPLAY_NAME];
+      const error = displayName?.length > 20;
+      return error && "Display Name should be 20 characters or less";
```
Comment (kiram15) [inline src/components/settings/SettingsLMSTab/LMSConfigs/Canvas/CanvasConfigAuthorizePage.tsx:32]: nit: should be "Display **n**ame should .... "

File: src/components/forms/FormWorkflow.tsx (hunk 6 above)
Comment (marlonkeating) [reply src/components/forms/FormWorkflow.tsx:102]: We'd need to import the Button props from Paragon, and Paragon currently isn't exporting any.

File: src/components/forms/FormWorkflow.tsx (hunk 5 above)
Comment (marlonkeating) [reply src/components/forms/FormWorkflow.tsx]: It isn't needed, I'm pretty sure these were leftovers when I was debugging that portion of code.

File: src/components/settings/SettingsLMSTab/LMSConfigs/Canvas/CanvasConfigAuthorizePage.tsx [hunk 11]
```diff
@@ -0,0 +1,153 @@
...
+export const validations: FormFieldValidation[] = [
+  {
+    formFieldId: formFieldNames.CANVAS_BASE_URL,
+    validator: (fields) => {
+      const error = !urlValidation(fields[formFieldNames.CANVAS_BASE_URL]);
+      return error && "Please enter a valid URL";
+    },
+  },
+  {
+    formFieldId: formFieldNames.DISPLAY_NAME,
```
Comment (kiram15) [inline src/components/settings/SettingsLMSTab/LMSConfigs/Canvas/CanvasConfigAuthorizePage.tsx:41]: Something's going wrong and when I key smash to put in a display name, this is my error message. "canvasBaseUrlPlease enter a valid URL", so something's wrong with the variable name being appended to the validation message, but also the incorrect ones getting triggered.

Comment (marlonkeating) [reply src/components/settings/SettingsLMSTab/LMSConfigs/Canvas/CanvasConfigAuthorizePage.tsx:44]: Good call out that it wasn't following the old behavior!

**Review by kiram15 — APPROVED** [review]

**Pull request merged**